Content.Server/Mining/MineableSystem.cs

```diff
@@ -36,36 +40,57 @@ private void OnInteractUsing(EntityUid uid, MineableComponent component, Interac
         if (!pickaxe.MiningEntities.Add(uid))
             return;
 
-        var doAfter = new DoAfterEventArgs(args.User, component.BaseMineTime * pickaxe.MiningTimeMultiplier, default, uid)
+        component.CancelToken = new CancellationTokenSource();
+        var doAfter = new DoAfterEventArgs(args.User, component.BaseMineTime * pickaxe.MiningTimeMultiplier, component.CancelToken.Token, uid)
         {
             BreakOnDamage = true,
             BreakOnStun = true,
             BreakOnTargetMove = true,
             BreakOnUserMove = true,
             MovementThreshold = 0.5f,
-            BroadcastCancelledEvent = new MiningDoafterCancel() { Pickaxe = args.Used, Rock = uid },
-            BroadcastFinishedEvent = new MiningDoafterSuccess() { Pickaxe = args.Used, Rock = uid }
+            BroadcastCancelledEvent = new MiningDoafterCancel(component) { Pickaxe = args.Used, Rock = uid },
+            TargetFinishedEvent = new MiningDoafterSuccess() { Pickaxe = args.Used, Rock = uid, Player = args.User }
         };
 
         _doAfterSystem.DoAfter(doAfter);
     }
 
-    private void OnDoafterSuccess(MiningDoafterSuccess ev)
+    private void OnDoafterSuccess(EntityUid uid, MineableComponent component, MiningDoafterSuccess ev)
     {
+        var entityManager = IoCManager.Resolve<IEntityManager>();
```

Comment:
EntityManager is already a property on systems.

Reply:
solved